File: src/main/java/algo/depthFirstSearch/WildcardMatching_44.java
package algo.depthFirstSearch;

public class WildcardMatching_44 {
  public static boolean isMatch(String s, String p) {
    // remove consecutive * from p
    var pattern = "";
    for (var c : p.toCharArray()) {
      if (!pattern.isEmpty() && c == '*' && pattern.charAt(pattern.length() - 1) == '*') {
        continue;
      }
      pattern += c;
    }

    return match(s, pattern, 0, 0, new int[s.length()][p.length()]) == 1;
  }

  private static int match(String s, String p, int sIndex, int pIndex, int[][] dp) {
    if (s.isEmpty()) {
      return "*".equals(p) || p.isEmpty() ? 1 : -1;
    }
    if (p.isEmpty()) {
      return -1;
    }

    int slen = s.length(), plen = p.length();
    // return previously calculated result
    if (dp[sIndex][pIndex] != 0) {
      return dp[sIndex][pIndex];
    }
    int i, j;
    for (i = sIndex, j = pIndex; i < slen && j < plen; i++, j++) {
      char schar = s.charAt(i), pchar = p.charAt(j);
      if (pchar == '*') {
        if (j == p.length() - 1) {
          // last '*' matches all
          dp[sIndex][pIndex] = 1;
          return 1;
        } else {
          for (var k = i; k < slen; k++) {
            if (s.charAt(k) == p.charAt(j + 1) || p.charAt(j + 1) == '?') {
              var result = match(s, p, k, j + 1, dp);
              dp[k][j + 1] = result;
              if (result == 1) {
                return 1;
              }
            }
          }
        }
      }
      if (schar != pchar && pchar != '?') {
        dp[sIndex][pIndex] = -1;
        return -1;
      }
    }
    // ignore last '*'
    if (i == slen && (j == plen || j == plen - 1 && p.charAt(j) == '*')) {
      dp[sIndex][pIndex] = 1;
    } else {
      dp[sIndex][pIndex] = -1;
    }
    return dp[sIndex][pIndex];
  }

  public static void main(String[] args) {
    System.out.println(isMatch("aa", "a"));
    System.out.println(isMatch("aa", "*"));
    System.out.println(isMatch("cb", "?a"));
    System.out.println(isMatch("aa", "aaa"));
    System.out.println(isMatch("aabbcc", "a?*b"));
    System.out.println(isMatch("adceb", "*a*b"));
    System.out.println(isMatch("adceb", "*a*b*"));
    System.out.println(isMatch("adceb", "*a*e*"));
    System.out.println(isMatch("", "*"));
    System.out.println(isMatch("", ""));
    System.out.println(isMatch("a", "*?*"));
//    System.out.println(isMatch2("abbabaaabbabbaababbabbbbbabbbabbbabaaaaababababbbabababaabbababaabbbbbbaaaabababbbaabbbbaabbbbababababbaabbaababaabbbababababbbbaaabbbbbabaaaabbababbbbaababaabbababbbbbababbbabaaaaaaaabbbbbaabaaababaaaabb",
//            "**aa*****ba*a*bb**aa*ab****a*aaaaaa***a*aaaa**bbabb*b*b**aaaaaaaaa*a********ba*bbb***a*ba*bb*bb**a*b*bb"));
  }
}

File: src/main/kotlin/algo/depthFirstSearch/WildcardMatching_44.kt
package algo.depthFirstSearch

import kotlin.contracts.contract

class KtWildcardMatching_44 {
    fun isMatch(s: String, p: String): Boolean {
        fun match(s: String, p: String, sIndex: Int, pIndex: Int, dp: Array<IntArray>): Int {
            if (s == "") {
                return if (p == "*" || p == "") 1 else 0
            }
            if (p == "") {
                return -1
            }
            if (dp[sIndex][pIndex] != 0) {
                return dp[sIndex][pIndex]
            }
            val sLen = s.length; val pLen = p.length
            var i = sIndex; var j = pIndex
            while (i < sLen && j < pLen) {
                val sChar = s[i]; val pChar = p[j]
                if (pChar == '*') {
                    if (j == pLen - 1) {
                        // pattern ends with '*' so it matches from current till end
                        dp[sIndex][pIndex] = 1
                        return 1
                    } else {
                        for (k in i..<sLen) {
                            // match the char after '*'
                            if (s[k] == p[j + 1] || p[j + 1] == '?') {
                                val res = match(s, p, k, j + 1, dp)
                                dp[k][j + 1] = res
                                if (res == 1) {
                                    return 1
                                }
                            }
                        }
                    }
                }
                if (pChar != '?' && sChar != pChar) {
                    dp[sIndex][pIndex] = -1
                    return -1
                }
                i++
                j++
            }
            if (i == sLen && (j == pLen || j == pLen - 1 && p[j] == '*')) {
                dp[sIndex][pIndex] = 1
            } else {
                dp[sIndex][pIndex] = -1
            }
            return dp[sIndex][pIndex]
        }

        // remove consecutive '*' in pattern
        var pattern = ""
        p.indices.forEach { i ->
            if (i == 0 || p[i] != '*' || p[i] == '*' && p[i - 1] != '*') {
                pattern += p[i]
            }
        }
        return match(s, pattern, 0, 0, Array(s.length) { IntArray(p.length) {0} }) == 1
    }
}

fun main() {
    val match = KtWildcardMatching_44()
    println(match.isMatch("aa", "a"))
    println(match.isMatch("aa", "*"))
    println(match.isMatch("cb", "?a"))
    println(match.isMatch("aa", "aaa"))
    println(match.isMatch("aabbcc", "a?*b"))
    println(match.isMatch("adceb", "*a*b"))
}